accepted/system-range/system-range.md

```diff
@@ -0,0 +1,226 @@
+# System.Range
+
+**PM** [Immo landwerth](https://github.com/terrajobst)
+
+The C# team is considering providing language syntax for specifying ranges to
+represent parts of a string, array, or span. In order to make this work, the
+language has to be bound to a .NET type, which we call `System.Range`.
+
+## Scenarios and User Experience
+
+### Construcing ranges
+
+```csharp
+// Ranges are constructed with an index and a length
+
+var fiveToTen = new Range(index: 5, length: 6);
+
+// By using a factory method, they can also be constructed from two indices.
+
+var alsoFiveToTen = Range.Between(5, 11);
+
+// The upper bound is considered exclusive so this will print the 1, 2, 3
+
+foreach (var value in Range.Between(1, 4))
+    Console.WriteLine(value);
+
+```
+
+The C# syntax for ranges isn't finalized and is actively worked on. While this
+is part of a separate design group it makes sense to introduce the syntax here
+so that we can look at source code as it's likely written:
+
+```csharp
+var fiveToTen = 5..11; // Equivalent to Range.Between(5, 11)
+```
+
+### Open ended ranges
+
+```csharp
+// Ranges can be open ended:
+
+var fiveToEnd = 5..;  // Equivalent to Range.From(5) i.e. missing upper bound
+var startToTen = ..1; // Equivalent to Range.To(11) i.e. missing lower bound
+var everything = ..;  // Equivalent to Range.All i.e. missing upper and lower bound
+
+// Asking for Index, Length or enumerating an open range will throw
+// NotSupportedException. The consumer of the range is expected to 
+// close it over the particular data:
+
+void Print(int[] data, Range range)
+{
+    if (range.IsOpen)
+        range = range.Close(0..range.Length);
```

Comment:
Personally I find this awkward.  What's the reasoning for not allowing users to query if Index or Length is set?

There is also a bug in your code.  From the comments `// Asking for Index, Length or enumerating an open range will throw`, however you are accessing `range.Length` inside of the `if (range.IsOpen)` check.

Reply:
> There is also a bug in your code.

Oops, the code is meant to read as follows:

```csharp
void Print(int[] data, Range range)
{
    if (range.IsOpen)
        range = range.Close(0..data.Length);
```

> What's the reasoning for not allowing consumers to query if `Index` or `Length` is set? 

Originally that's what I did but then I got feedback that it makes the type more busy. How would you model it? The most natural representation would be `int?` but that makes writing checks more awkward. It also means that that callers would have to check for both ends when in practice I think you don't care which end is open, you just want to close it. Partially closing seems like a non-scenario to me (although you can if `Close` accepts a range).